File: src/core/summarizer.py
"""The Summarizer Worker - compresses narrative memory."""

from ..models.session import GameSession
from ..prompts.builder import PromptBuilder
from .gemini import GeminiClient


class SummarizerWorker:
    """Handles memory compression every 5 turns.

    The Librarian takes existing long-term memory + recent events
    and produces a condensed summary that preserves critical information
    while staying within token limits.
    """

    # Maximum words for long-term memory
    MAX_MEMORY_WORDS = 200

    # Trigger threshold
    TURN_THRESHOLD = 5

    def __init__(self, client: GeminiClient):
        self.client = client

    def should_run(self, session: GameSession) -> bool:
        """Check if summarization is needed."""
        return session.needs_summary()

    def run(self, session: GameSession) -> str:
        """Execute summarization and update session.

        Returns:
            The new compressed memory string
        """
        if not session.narrative:
            return ""

        # Gather recent turns for compression
        recent_turns = [
            {"role": turn.role, "text": turn.text}
            for turn in session.narrative.short_term_buffer
        ]

        if not recent_turns:
            return session.narrative.long_term_memory

        # Build summarization prompt
        builder = PromptBuilder(session)
        summary_prompt = builder.build_summary_prompt(recent_turns)

        # Generate new summary
        new_memory = self.client.generate_summary(summary_prompt)

        # Enforce word limit
        words = new_memory.split()
        if len(words) > self.MAX_MEMORY_WORDS:
            # Recursive summarization - summarize the summary
            new_memory = self._recursive_compress(new_memory)

        # Update session
        session.narrative.long_term_memory = new_memory
        session.mark_summarized()

        return new_memory

    def _recursive_compress(self, memory: str) -> str:
        """Emergency compression when memory exceeds limits."""
        compress_prompt = f"""URGENT: The following summary is too long. Compress it to under 150 words while keeping ALL [CLUE] markers and critical facts.

CURRENT SUMMARY:
{memory}

Respond with ONLY the compressed summary."""

        compressed = self.client.generate_summary(compress_prompt)

        # Hard truncation as last resort
        words = compressed.split()
        if len(words) > self.MAX_MEMORY_WORDS:
            compressed = " ".join(words[:self.MAX_MEMORY_WORDS]) + "..."

        return compressed

    def force_checkpoint(self, session: GameSession) -> str:
        """Force an immediate summary regardless of turn count.

        Useful for critical moments or before potential game-ending events.
        """
        return self.run(session)

File: src/core/summarizer.ts
/**
 * The Summarizer Worker - compresses narrative memory.
 */

import { GameSession, needsSummary, markSummarized } from '../models/session.js';
import { Turn } from '../models/narrative.js';
import { buildSummaryPrompt } from '../prompts/builder.js';
import { GeminiClient } from './gemini.js';

/**
 * Maximum words for long-term memory.
 */
const MAX_MEMORY_WORDS = 200;

/**
 * Handles memory compression every 5 turns.
 */
export class SummarizerWorker {
  constructor(private client: GeminiClient) {}

  /**
   * Check if summarization is needed.
   */
  shouldRun(session: GameSession): boolean {
    return needsSummary(session);
  }

  /**
   * Execute summarization and update session.
   */
  async run(session: GameSession): Promise<string> {
    if (!session.narrative) {
      return '';
    }

    const recentTurns = session.narrative.shortTermBuffer;
    if (recentTurns.length === 0) {
      return session.narrative.longTermMemory;
    }

    // Build summarization prompt
    const summaryPrompt = buildSummaryPrompt(
      session.narrative.longTermMemory,
      recentTurns
    );

    // Generate new summary
    let newMemory = await this.client.generateSummary(summaryPrompt);

    // Enforce word limit
    const words = newMemory.split(/\s+/);
    if (words.length > MAX_MEMORY_WORDS) {
      newMemory = await this.recursiveCompress(newMemory);
    }

    // Update session
    session.narrative.longTermMemory = newMemory;
    markSummarized(session);

    return newMemory;
  }

  /**
   * Emergency compression when memory exceeds limits.
   */
  private async recursiveCompress(memory: string): Promise<string> {
    const compressPrompt = `URGENT: The following summary is too long. Compress it to under 150 words while keeping ALL [CLUE] markers and critical facts.

CURRENT SUMMARY:
${memory}

Respond with ONLY the compressed summary.`;

    let compressed = await this.client.generateSummary(compressPrompt);

    // Hard truncation as last resort
    const words = compressed.split(/\s+/);
    if (words.length > MAX_MEMORY_WORDS) {
      compressed = words.slice(0, MAX_MEMORY_WORDS).join(' ') + '...';
    }

    return compressed;
  }

  /**
   * Force an immediate summary regardless of turn count.
   */
  async forceCheckpoint(session: GameSession): Promise<string> {
    return this.run(session);
  }
}
